From c295af47bdd556e4b0966c9adf4a882f7b651c7b Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Thu, 6 Apr 2017 14:09:44 +0800
Subject: [PATCH 1656/5242] MLK-14630 video: mipi_dsi_northwest: refine
 suspend logic

commit  ae87d8c53be4306871e9be69918df72f15e00733 from
https://source.codeaurora.org/external/imx/linux-imx.git

Move the 'lcd_inited' flag status modification from
'mipi_dsi_disable' to 'mipi_dsi_suspend' to make sure
this flag to be set to 0 during system suspend.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit 02d805b48bfb1fc637f9ccce4ed7ae59d62e7d2f)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/mipi_dsi_northwest.c |   20 +++++++++++++-------
 1 file changed, 13 insertions(+), 7 deletions(-)

diff --git a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
index 2af7186..9ff5d92 100644
--- a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
+++ b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
@@ -773,13 +773,6 @@ static void mipi_dsi_disable(struct mxc_dispdrv_handle *disp,
 	if (!mipi_dsi->encoder)
 		mipi_display_enter_sleep(mipi_dsi->disp_mipi);
 
-	if (fbi->state == FBINFO_STATE_SUSPENDED) {
-		writel(0x1, mipi_dsi->mmio_base + DPHY_PD_PLL);
-		writel(0x1, mipi_dsi->mmio_base + DPHY_PD_DPHY);
-
-		mipi_dsi->lcd_inited = 0;
-	}
-
 	clk_disable_unprepare(mipi_dsi->esc_clk);
 
 	reset_dsi_domains(mipi_dsi, 1);
@@ -1042,6 +1035,19 @@ static int mipi_dsi_runtime_resume(struct device *dev)
 
 static int mipi_dsi_suspend(struct device *dev)
 {
+	struct platform_device *pdev = to_platform_device(dev);
+	struct mipi_dsi_info *mipi_dsi = dev_get_drvdata(&pdev->dev);
+
+	if (unlikely(mipi_dsi->lcd_inited)) {
+		clk_prepare_enable(mipi_dsi->esc_clk);
+
+		writel(0x1, mipi_dsi->mmio_base + DPHY_PD_PLL);
+		writel(0x1, mipi_dsi->mmio_base + DPHY_PD_DPHY);
+
+		clk_disable_unprepare(mipi_dsi->esc_clk);
+		mipi_dsi->lcd_inited = 0;
+	}
+
 	pinctrl_pm_select_sleep_state(dev);
 
 	return 0;
-- 
1.7.9.5

